File: lintcode/gray-code.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/gray-code/

class Solution:
    # @param {int} n a number
    # @return {int[]} Gray code
    def grayCode(self, n):
        # Write your code here

        result = [0]
        for i in xrange(n):
            count = len(result)
            for j in xrange(count - 1, -1, -1):
                result.append(result[j] | pow(2, i))
        return result

if __name__ == '__main__':
    s = Solution()
    print s.grayCode(0)
    print s.grayCode(1)
    print s.grayCode(2)
    print s.grayCode(3)
    print s.grayCode(4)

File: leetcode-2/gray-code.cc
// Question: https://leetcode.com/problems/gray-code/

class Solution {
public:
    vector<int> grayCode(int n) {
        vector<int> ans;
        ans.push_back(0);
        if (n == 0) {
            return ans;
        }

        ans.push_back(1);
        for (int i = 1; i < n; ++i) {
            int count = ans.size();
            for (int j = count-1; j >= 0; --j) {
                ans.push_back(ans.at(j) | (1 << i));
            }
        }

        return ans;
    }
};
